### PR TITLE
feat: add public key fetcher and signature verifier

### PR DESCRIPTION
**Pull request by vincentdelmundo**
# Description
Describe the changes in this pull request.
Explain your rationale of technical decisions you made (unless discussed before).

## Links
Add links to github/jira issues, design documents and other relevant resources (e.g. previous discussions, platform/tool documentation etc.)

# Checklist
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md)
- [x] I have not committed any secrets (passwords, tokens, internal urls etc.) or sensitive information
- [x] I wrote/updated tests for new/changed code
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

**Review by leila-gu — APPROVED** [review]
LGTM

**Comment by leila-gu** [comment]
Since the repo is using Codecov, it's nice to have the readme icon too!

File: signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    companion object {
+        private const val KEYSTORE_ALIAS = "signature-verifier-public-key-encryption-decryption"
+    }
+}
+
+internal class PreferenceEncryptorKeyGenerator(private val alias: String) {
+    fun generateKey(context: Context) =
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val spec = KeyGenParameterSpec.Builder(
```
Comment (corycaywood) [inline signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt:38]: Is there any advantage to manually defining a `KeyGenParameterSpec` vs using a default like:
`MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()`

**Review by corycaywood — COMMENTED** [review]
Looking good. I think we just need to do something about the `MasterKey` issue first.

Also, will you be adding some test functionality in the `sample` app?

File: signatureverifier/build.gradle [hunk 2]
```diff
@@ -29,14 +31,26 @@ android {
...
-
     implementation "org.jetbrains.kotlin:kotlin-stdlib:$CONFIG.versions.kotlin"
+    implementation "com.squareup.okhttp3:okhttp:$CONFIG.versions.okhttp"
+    implementation "com.squareup.moshi:moshi:$CONFIG.versions.moshi"
+    kapt "com.squareup.moshi:moshi-kotlin-codegen:$CONFIG.versions.moshi"
     implementation 'androidx.core:core-ktx:1.3.2'
-    implementation 'androidx.appcompat:appcompat:1.2.0'
-    implementation 'com.google.android.material:material:1.3.0'
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.9"
+    implementation "androidx.security:security-crypto:1.1.0-alpha03"
```
[the remark this answers is not in the document]
Comment (vincentdelmundo) [reply signatureverifier/build.gradle:40]: Changed to compatible version

File: signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                alias,
+                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
+            )
+                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+                .setKeySize(MasterKey.DEFAULT_AES_GCM_MASTER_KEY_SIZE)
+                .build()
+            MasterKey.Builder(context).setKeyGenParameterSpec(spec).build()
```
[the remark this answers is not in the document]
Comment (vincentdelmundo) [reply signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt:46]: added try/catch for valid MasterKey creation

File: signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt (hunk 1 above)
Comment (vincentdelmundo) [reply signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt:38]: The only difference is the specs/settings is more customizable when using `KeyGenParameterSpec`. But you're right, using `setKeyScheme` is enough.
Updated this.

File: signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    fun getEncryptedKey(keyId: String) = sharedPref.getString(keyId, null)
+
+    companion object {
+        private const val KEYSTORE_ALIAS = "signature-verifier-public-key-encryption-decryption"
+    }
+}
+
+internal class PreferenceEncryptorKeyGenerator(private val alias: String) {
+    fun generateKey(context: Context) =
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
[the remark this answers is not in the document]
Comment (vincentdelmundo) [reply signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt:37]: Yeah, need to discuss this.
For now, the SignatureVerifier SDK use the min SDK set in `buildconfig` (i.e. 21)

File: signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                alias,
+                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
+            )
+                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+                .setKeySize(MasterKey.DEFAULT_AES_GCM_MASTER_KEY_SIZE)
+                .build()
+            MasterKey.Builder(context).setKeyGenParameterSpec(spec).build()
+        } else {
+            MasterKey.Builder(context).build()
```
[the remark this answers is not in the document]
Comment (vincentdelmundo) [reply signatureverifier/src/main/kotlin/io/github/rakutentech/signatureverifier/verification/SharedPreferenceEncryptor.kt:48]: This was updated based on another comment (added `KeyScheme`)

**Comment by vincentdelmundo** [comment]
> Looking good. I think we just need to do something about the `MasterKey` issue first.
> 
> Also, will you be adding some test functionality in the `sample` app?

Yes, when the endpoint is ready (at least in staging), sample app will be updated and will add in react-native demoapp.

**Review by corycaywood — APPROVED** [review]
LGTM 👍

**Pull request merged**